### ISSUE TITLE
[Performance] Upgrade from the legacy `DrawerNavigator` to the modern one that uses Reanimated 2

### ISSUE THREAD

**Issue by marcaaron** [Expensify/App#11426]
# Problem

- We are still using the legacy version of the Drawer Navigator which is not using Reanimated 2.

- I discovered that the DrawerView is not correctly picking up that Reanimated 2 is configured so it falls back to the "legacy" version of the drawer. The new code we want [here](https://github.com/react-navigation/react-navigation/blob/e13b4d9341362512ba4bf921a17552f3be8735c1/packages/drawer/src/views/DrawerView.tsx#L83-L95) is in a newer version of the [`@react-navigation/drawer`](https://github.com/react-navigation/react-navigation/blob/main/packages/drawer) package.

- Upgrading will give massive improvements on Android to navigating and will allow us to use the `InteractionManager` to properly defer some expensive things we are doing until after the drawer opens/closes

HOWEVER...

There are two major blockers for this upgrade.

1. When the drawer performance is improved on Android it becomes so good that you see the previous chat before we display the new chat. Without the improvement today things just hang (for a few seconds) and we navigate to the new report. We will need to solve this somehow. Perhaps by showing the loader overlay as soon as a chat is tapped

2. Something about this upgrade is breaking the behavior of the drawer on mobile web. It will not open or close or respond to `DrawerActions` at all. We cannot fallback to the `useLegacyImplementation` as we'll get this error:

![Screen Shot 2022-09-29 at 7 00 50 AM](https://user-images.githubusercontent.com/32969087/193093359-9538f98e-0166-48fd-ac4e-7deaa8c20717.png)

If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by marcaaron** [comment]
I've been looking into part 2 of this and pretty confused about why the drawer transitions are breaking in App specifically. I looked at the example app in the react navigation repo and ran it locally with settings similar to ours and it works OK so doesn't seem like there is any obvious regression.

While inspecting it just looks like the html elements are not updating at all with the new styles.

We _should_ be seeing values update like this on `main`:

https://user-images.githubusercontent.com/32969087/193312104-7fdfe1df-521d-43e4-986c-756d672833e9.mp4

But instead the `translateX` values are not updating at all 🤔

**Comment by marcaaron** [comment]
test branch: https://github.com/Expensify/App/compare/marcaaron-enableNonLegacyDrawer

**Comment by Szymon20000** [comment]
Are you sure that the issue with using legacy drawer is also present on Android?
I would suspect it's only problem on web based on changes here.

**Comment by marcaaron** [comment]
No to be clear, Android works great! It's _only_ a problem on web. But cannot seem to downgrade to the legacy drawer on web. I tried to do some debugging in the `Drawer` component and think it might be an issue with `useAnimatedStyle`, but it's the first time I've look at Reanimated so not 100% sure.

**Comment by Szymon20000** [comment]
I found the issue and responded on slack

**Comment by Szymon20000** [comment]
https://expensify.slack.com/files/U0395FWE6TZ/F044R9C3B7V/screenshot_2022-10-01_at_17.15.58.png

**Comment by marcaaron** [comment]
**Update:** Still working on this. Mainly have been blocked on a change that the upgrade caused for the native side. Before the upgrade, tapping on a drawer item would make the UI freeze up until the next chat screen is ready to view. After the upgrade we navigate immediately, but still the old chat can be seen before transitioning to the new chat. 

I think both are downsides, but I'd rather have the drawer tap be responsive and then wait some time before showing the next chat. I tried experimenting with showing a loading state during the transition, but seems like the native view takes some time to drop. This was most obvious on Android, but can be seen on iOS as well.

The latest thing I tried was wrapping some expensive things in `SidebarLinks` and `ReportScreen` with `<Freeze>` and setting the `freeze` prop to depend on the drawer state. The results are pretty good on an Android release build on emulator:

https://user-images.githubusercontent.com/32969087/194616753-f88749d5-e05c-4578-83d0-7fa3c3b612b8.mp4

The only downside left is that the frozen elements show as hidden when we are swiping. I think we are planning to move away from the drawer to a stack so not sure how much more energy should be put into perfecting this.

On the other hand, Android release build performance for chat switching is starting to look really good, I suspect from some other recent changes to Onyx and improvements in App:

https://user-images.githubusercontent.com/32969087/194618968-f33fce57-3d98-4f20-b965-823c8ee84008.mp4

So, it's hard to tell the difference now 😄 but the interaction is much more responsive and snappy in the first example with the upgraded `Drawer`.
